expression/builtin_vectorized.go

```diff
@@ -0,0 +1,207 @@
+	builtinFunc
+}
+
+func (c *vecRowConverter) vecEval(input *chunk.Chunk, result *chunk.Column) error {
+	if c.builtinFunc.vectorized() {
+		return c.builtinFunc.vecEval(input, result)
+	}
+
+	// convert from row-based evaluation to vectorized evaluation.
+	// evaluate each row in input according to its row-based evaluation methods and updates the result.
+	it := chunk.NewIterator4Chunk(input)
+	row := it.Begin()
+	var isNull bool
+	var err error
+	switch c.builtinFunc.getRetTp().EvalType() {
+	case types.ETInt:
+		result.ResizeInt64(input.NumRows())
+		i64s := result.Int64s()
+		for i := range i64s {
+			if i64s[i], isNull, err = c.builtinFunc.evalInt(row); err != nil {
+				return err
+			}
+			result.SetNull(i, isNull)
+			row = it.Next()
+		}
+	case types.ETReal:
+		result.ResizeFloat64(input.NumRows())
+		f64s := result.Float64s()
+		for i := range f64s {
+			if f64s[i], isNull, err = c.builtinFunc.evalReal(row); err != nil {
+				return err
+			}
+			result.SetNull(i, isNull)
+			row = it.Next()
+		}
+	case types.ETDecimal:
+		result.ResizeDecimal(input.NumRows())
+		ds := result.Decimals()
+		var v *types.MyDecimal
+		for i := range ds {
+			if v, isNull, err = c.builtinFunc.evalDecimal(row); err != nil {
+				return err
+			}
+			if isNull {
+				result.SetNull(i, true)
+			} else {
+				result.SetNull(i, false)
+				ds[i] = *v
+			}
+			row = it.Next()
+		}
+	case types.ETDuration:
+		result.ResizeDuration(input.NumRows())
+		ds := result.GoDurations()
+		var v types.Duration
+		for i := range ds {
+			if v, isNull, err = c.builtinFunc.evalDuration(row); err != nil {
+				return err
+			}
+			ds[i] = v.Duration
+			result.SetNull(i, isNull)
+			row = it.Next()
+		}
+	case types.ETJson:
+		result.ReserveJSON(input.NumRows())
+		var v json.BinaryJSON
+		for ; row != it.End(); row = it.Next() {
+			if v, isNull, err = c.builtinFunc.evalJSON(row); err != nil {
+				return err
+			}
+			if isNull {
+				result.AppendNull()
+			} else {
+				result.AppendJSON(v)
+			}
+		}
+	case types.ETString:
+		result.ReserveString(input.NumRows())
+		var v string
+		for ; row != it.End(); row = it.Next() {
+			if v, isNull, err = c.builtinFunc.evalString(row); err != nil {
+				return err
+			}
+			if isNull {
+				result.AppendNull()
+			} else {
+				result.AppendString(v)
+			}
+		}
+	case types.ETDatetime, types.ETTimestamp:
+		result.Reset()
+		var v types.Time
+		for ; row != it.End(); row = it.Next() {
+			if v, isNull, err = c.builtinFunc.evalTime(row); err != nil {
+				return err
+			}
+			if isNull {
+				result.AppendNull()
+			} else {
+				result.AppendTime(v)
+			}
+		}
+	default:
+		return errors.Errorf("unsupported type for converting from row-based to vectorized evaluation, please contact the TiDB team for help")
+	}
+	return nil
+}
+
+func (c *vecRowConverter) vectorized() bool {
+	return true
+}
+
+func (c *vecRowConverter) evalInt(row chunk.Row) (val int64, isNull bool, err error) {
+	if c.builtinFunc.vectorized() {
```

Comment:
If we make a `newVecRowConverter(builtin builtinFunc)`, and check `vectorized` there, then we don't need to check it in every eval method.

Reply:
And we don't even need to override eval methods.

Reply:
In the next PR, I will implement converting from vectorized evaluation to row-based evaluation, then we need to do this check.

Reply:
But you are right, I will remove these overriding in this PR.

Reply:
These unused methods have been removed from this PR.